database_api.py

```diff
@@ -50,9 +51,317 @@ def get_all_test(self):
+        # or AUDIT, NAUDIT (for AUDIT)
+        # Assume only one PDF option, one AUDIT option can be selected
+        PDF_MAPPINGS = {
+            "PDF": {"required": False, "permitted": True},
+            "PDFO": {"required": True, "permitted": True},
+            "NPDF": {"required": False, "permitted": False},
+        }
+
+        AUDIT_MAPPINGS = {"AUDIT": True, "NAUDIT": False}
+
+        if len(grading) > 0:
+            get_courses = True
+            pdf, audit = None, None
+            for code in grading:
+                if code in PDF_MAPPINGS.keys():
+                    pdf = PDF_MAPPINGS[code]
+                elif code in AUDIT_MAPPINGS.keys():
+                    audit = AUDIT_MAPPINGS[code]
+
+            if pdf is not None:
+                course_filters["$and"].append({"pdf": pdf})
+
+            if audit is not None:
+                course_filters["$and"].append({"audit": audit})
+
+        # Search by course code + number, title, or instructor
+        if string != "":
+            get_courses = True
+            course_filters["$and"].append(
+                {
+                    "$or": [
+                        {"title": {"$regex": string, "$options": "i"}},
+                        {
+                            "catalog_title": {
+                                "$regex": string.replace(" ", ""),
+                                "$options": "i",
+                            }
+                        },
+                    ]
+                }
+            )
+
+            names = string.split(" ")
+            names = list(
+                map(lambda name: {"$regex": r"^" + name, "$options": "i"}, names)
+            )
+            instr_filters["$or"] = []
+            for name in names:
+                instr_filters["$or"].extend(
+                    [{"name.first_name": name}, {"name.last_name": name},]
+                )
+
+        if not get_courses:
+            return []
+
+        # For debugging search tests:
+        # print("== Course Filters: ==")
+        # print(course_filters)
+        # print("== Instructor Filters: ==")
+        # print(instr_filters)
+
+        course_res = self.db.courses.find(course_filters)
+        instr_res = self.db.instructors.find(instr_filters)
+
+        # TO-DO: do sort after retrieving results
+        # TO-DO: return course information for each instructor retrieved
+
+        # result[0] is list of courses, result[1] is list of instructors
+        return [list(course_res), list(instr_res)]
+
 
 if __name__ == "__main__":
     # a basic example of how to use, can remove later
+    import copy
+
     db = DatabaseAPI()
+
+    """
+        query = {
+            "string": "the query string",
+            "semester": "one of the semester",
+            "sort": "Course Code, Relevance, Rating, Title",
+            "filters": {
+                "special": "*" or "NEW",
+                "dist": [list of dist codes],
```

Comment:
dist or dists?

Reply:
dists, sorry i changed the key names in between